Snpe ep example (#116)
microsoft/onnxruntime-inference-examples: c_cxx/Snpe_EP/README.md

# Image classification with Inception v3 in CPP using SNPE Execution Provider

> 1.  The imagel classification uses a Inception v3 model from Snpe turtorials with DLC format. Wrap it as a Onnx model in Onnx model to inference with SNPE Execution Provider.

Reviewer: insert link from SNPE page.
```suggestion
1.  This image classification sample uses the Inception v3 model from [SNPE turtorial](https://developer.qualcomm.com/sites/default/files/docs/snpe/tutorial_inceptionv3.html) with DLC format. Wrap it as a custom node in an ONNX model to inference with SNPE Execution Provider.

```

2.  The sample uses the Onnxruntime SNPE Execution Provider to run inference on various Qualcomm devices like Qualcomm CPU, GPU DSP, AIP, etc. It supports Windows ARM64, and Android.

# Prerequisites (Linux host)
1. Get a Linux host

> 2. Download SNPE SDK from [here](https://developer.qualcomm.com/software/qualcomm-neural-processing-sdk)

Reviewer: ```suggestion
2. Download SNPE SDK from Qualcomm's developer site [here](https://developer.qualcomm.com/software/qualcomm-neural-processing-sdk)

```

3. Setup SNPE on the Linux host. Setup environment for the tutorial. Follow the Tutorials and Examples for [Tensorflow inceptionv3](https://developer.qualcomm.com/sites/default/files/docs/snpe/tutorial_inceptionv3.html)
4. Get the model [Inception v3](https://developer.qualcomm.com/sites/default/files/docs/snpe/tutorial_setup.html#tutorial_setup_inception_v3)

    ```
    python $SNPE_ROOT/models/inception_v3/scripts/setup_inceptionv3.py -a ~/tmpdir -d -r dsp
    ```

    The generated DLC file inception_v3.dlc and inception_v3_quantized.dlc can be found at $SNPE_ROOT/models/inception_v3/dlc/.

    The data chairs.raw, notice_sign.raw, plastic_cup.raw and trash_bin.raw can be found at $SNPE_ROOT/models/inception_v3/data/. The sample applicatioin use these raw file as input.

> 5. Create Onnx model from DLC file

Reviewer: ```suggestion
5. Create ONNX model from DLC file

```

    Create a script gen_onnx_model.py in folder $SNPE_ROOT/models/inception_v3 with the code below:

    ```
    import onnx
    from onnx import helper
    from onnx import TensorProto

    with open('./dlc/inception_v3_quantized.dlc','rb') as file:
        file_content = file.read()

    input1 = helper.make_tensor_value_info('input:0', TensorProto.FLOAT, [1, 299, 299, 3])
    output1 = helper.make_tensor_value_info('InceptionV3/Predictions/Reshape_1:0', TensorProto.FLOAT, [1, 1001])

    snpe_node = helper.make_node('Snpe', name='Inception v3', inputs=['input:0'], outputs=['InceptionV3/Predictions/Reshape_1:0'], DLC=file_content, snpe_version='1.61.0', target_device='DSP', notes='quantized dlc model.', domain='com.microsoft')

    graph_def = helper.make_graph([snpe_node], 'Inception_v3', [input1], [output1])
    model_def = helper.make_model(graph_def, producer_name='tesorflow', opset_imports=[helper.make_opsetid('', 13)])
    onnx.save(model_def, 'snpe_inception_v3.onnx')
    ```

    Run the script to generate the Onnx model with the DLC content embed in a Onnx node.

>     Change the data type for input & output from TensorProto.FLOAT to TensorProto.UINT8 if you want to run model inference with quantized data. It's helpful to save the bandwidth (expecially for application with CS mode, data need to be transmitted across the network). Also need to change the options_values for buffer_type from FLOAT to TF8 in the main.cpp

Reviewer: Is the choice for the data type required to match the `provider_options` in the EP?

Author: that's what I was trying to say, let me make it clear.

2.  The sample uses the Onnxruntime SNPE Execution Provider to run inference on various Qualcomm devices like Qualcomm CPU, GPU DSP, AIP, etc. It supports Windows ARM64, and Android.

# Prerequisites (Linux host)
1. Get a Linux host
2. Download SNPE SDK from Qualcomm's developer site [here](https://developer.qualcomm.com/software/qualcomm-neural-processing-sdk)

3. Setup SNPE on the Linux host. Setup environment for the tutorial. Follow the Tutorials and Examples for [Tensorflow inceptionv3](https://developer.qualcomm.com/sites/default/files/docs/snpe/tutorial_inceptionv3.html)
4. Get the model [Inception v3](https://developer.qualcomm.com/sites/default/files/docs/snpe/tutorial_setup.html#tutorial_setup_inception_v3)

    ```
    python $SNPE_ROOT/models/inception_v3/scripts/setup_inceptionv3.py -a ~/tmpdir -d -r dsp
    ```

    The generated DLC file inception_v3.dlc and inception_v3_quantized.dlc can be found at $SNPE_ROOT/models/inception_v3/dlc/.

    The data chairs.raw, notice_sign.raw, plastic_cup.raw and trash_bin.raw can be found at $SNPE_ROOT/models/inception_v3/data/. The sample applicatioin use these raw file as input.

5. Create ONNX model from DLC file

>     Create a script gen_onnx_model.py in folder $SNPE_ROOT/models/inception_v3 with the code below:

Reviewer 2: it would be great to create a generic version of this simple script and check into main repo. keep it simple and have it accept only a few params such as input/output name/type/shape and input/output model file names

Author: Is it worth to do that much? To make it generic we need to think about more, like support multi-input, multi-output. so, need to have a list for input names and shapes as parameter which may end up not easy to use, and make the code complicated. I think this script is simple enough and straight forward for user to make modifications for their model.
